### ISSUE TITLE
Upgrading from 0.8.3 to 0.9.4 causes crashes

### ISSUE THREAD

**Issue by msimerson** [vstakhov/rspamd#288]
### From /var/log/messages:

````
May  9 15:01:01 rspamd pkg: rspamd-0.8.3 installed
May 25 22:07:43 rspamd pkg: pkg upgraded: 1.5.1 -> 1.5.2 
May 25 22:08:05 rspamd pkg: perl5 upgraded: 5.18.4_14 -> 5.20.2_4 
May 25 22:08:06 rspamd pkg: pcre upgraded: 8.35_2 -> 8.37 
May 25 22:08:06 rspamd pkg: glib reinstalled: 2.42.2 -> 2.42.2 
May 25 22:08:06 rspamd pkg: sqlite3 upgraded: 3.8.9_1 -> 3.8.10.1 
May 25 22:08:06 rspamd pkg: luajit-2.0.3_1 installed
May 25 22:08:06 rspamd pkg: hiredis-0.13.1 installed
May 25 22:08:08 rspamd pkg: rspamd upgraded: 0.8.3 -> 0.9.4
````

### Host info:

````
FreeBSD rspamd.tnpi.net 10.1-RELEASE-p9 FreeBSD 10.1-RELEASE-p9 #0: Tue Apr  7 01:09:46 UTC 2015     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64
````

### New and undesirable results

````
2015-05-26 11:16:44 #77355(main) main: normal process 25281 terminated abnormally by signal: 10
2015-05-26 11:16:54 #77355(main) main: normal process 32246 terminated abnormally by signal: 10
2015-05-26 11:19:23 #77355(main) main: normal process 35599 terminated abnormally by signal: 10
2015-05-26 11:19:40 #77355(main) main: normal process 32358 terminated abnormally by signal: 10
2015-05-26 11:20:03 #77355(main) main: normal process 32143 terminated abnormally by signal: 10
2015-05-26 11:20:21 #77355(main) main: normal process 35758 terminated abnormally by signal: 10
2015-05-26 11:20:40 #77355(main) main: normal process 25125 terminated abnormally by signal: 10
2015-05-26 11:21:29 #77355(main) main: normal process 25263 terminated abnormally by signal: 10
2015-05-26 11:21:36 #77355(main) main: normal process 38592 terminated abnormally by signal: 10
2015-05-26 11:22:08 #77355(main) main: normal process 38962 terminated abnormally by signal: 10
2015-05-26 11:22:36 #77355(main) main: normal process 36219 terminated abnormally by signal: 10
2015-05-26 11:22:42 #77355(main) main: normal process 32623 terminated abnormally by signal: 10
2015-05-26 11:23:23 #77355(main) main: normal process 39514 terminated abnormally by signal: 10
2015-05-26 11:23:58 #77355(main) main: normal process 77359 terminated abnormally by signal: 10
2015-05-26 11:24:00 #77355(main) main: normal process 36176 terminated abnormally by signal: 10
2015-05-26 11:26:04 #77355(main) main: normal process 39018 terminated abnormally by signal: 10
2015-05-26 11:26:37 #77355(main) main: normal process 39227 terminated abnormally by signal: 10
2015-05-26 11:26:39 #77355(main) main: normal process 35681 terminated abnormally by signal: 10
2015-05-26 11:27:08 #77355(main) main: normal process 32177 terminated abnormally by signal: 10
2015-05-26 11:27:13 #77355(main) main: normal process 39722 terminated abnormally by signal: 10
````

After initially installing via FreeBSD packages (pkg upgrade), I also rebuilt the package locally, with no change in behavior. The rspamd client (Haraka plugin) is reporting:

````
May 26 11:16:26 [rspamd] query failed: socket hang up
May 26 11:16:44 [rspamd] query failed: socket hang up
May 26 11:16:54 [rspamd] query failed: socket hang up
May 26 11:19:23 [rspamd] query failed: socket hang up
May 26 11:19:40 [rspamd] query failed: read ECONNRESET
May 26 11:20:03 [rspamd] query failed: socket hang up
May 26 11:20:21 [rspamd] query failed: socket hang up
May 26 11:20:40 [rspamd] query failed: write EPIPE
May 26 11:21:29 [rspamd] query failed: socket hang up
May 26 11:21:36 [rspamd] query failed: write ECONNRESET
````

Do you have suggestions for troubleshooting further?

**Comment by vstakhov** [comment]
Could you please make a backtrace from some core file produced by rspamd? I have some users who run rspamd on FreeBSD but I've not observed such a weird behaviour.

**Comment by msimerson** [comment]
> Could you please make a backtrace from some core file produced by rspamd?

I'd be happy too, but I'm not skilled at such things. `find / -name '*.core'` doesn't turn up anything, so apparently it's not dumping core. How do I get it to do so?  After having achieved that, how do I make a backtrace?

**Comment by baudehlo** [comment]
Matt check ulimit -c - it's usually set to 0 which prevents core dumps.

Then load in gdb (gdb -c <corefile>) and type "bt" and copy-paste that.

**Comment by msimerson** [comment]
```sh
#	ulimit -c
unlimited
```

That's set as the default (only entry) in /etc/login.conf.

**Comment by msimerson** [comment]
In case it helps, the log entries emitted by rspamd when starting up:

````
2015-05-26 13:40:48 #43241(main) register_symbol_common: symbol mailspike_wl is not registered in any metric, so skip its check
2015-05-26 13:40:48 #43241(main) register_symbol_common: symbol mailspike_bl is not registered in any metric, so skip its check
2015-05-26 13:40:48 #43241(main) ratelimit.lua:379: no servers are specified
2015-05-26 13:40:48 #43243(controller) rspamd_controller_password_sane: enable password is not set, so you should filter controller availability by using of firewall or `secure_ip` option
````

**Comment by vstakhov** [comment]
For FreeBSD I'd recommend the following steps (assuming `#` as root prompt):

```
# mkdir /coredump
# chmod 1777 /coredump
# sysctl kern.corefile=/coredump/%N-%P.core
# sysctl kern.coredump=1
# sysctl kern.sugid_coredump
```

Afterwards, build rspamd from the ports to preserve debug symbols:

```
# cd /usr/ports/mail/rspamd
# make package DEBUG_FLAGS="-g -O0" STRIP=
# make deinstall reinstall
```

Then you'd have core files inside `/coredump/rspamd-<pid>.core`
To get trace you might want to use `lldb` on FreeBSD 10:

```
# lldb /usr/local/bin/rspamd -c /coreland/rspamd-<pid>.core
...
lldb> bt
```

I'm sorry but that's quite nontrivial I know :(

**Comment by msimerson** [comment]
For reference, this is how to [install lldb](https://wiki.freebsd.org/lldb)

**Comment by msimerson** [comment]
````
root@rspamd # gdb -c /coredump/rspamd-0.9.4-58167.core
GNU gdb 6.1.1 [FreeBSD]
Copyright 2004 Free Software Foundation, Inc.
GDB is free software, covered by the GNU General Public License, and you are
welcome to change it and/or distribute copies of it under certain conditions.
Type "show copying" to see the conditions.
There is absolutely no warranty for GDB.  Type "show warranty" for details.
This GDB was configured as "amd64-marcel-freebsd".
Core was generated by `rspamd-0.9.4'.
Program terminated with signal 10, Bus error.
#0  0x0000000803566042 in ?? ()
(gdb) bt
#0  0x0000000803566042 in ?? ()
#1  0x0000000803867c40 in ?? ()
#2  0x00000008044000c0 in ?? ()
#3  0x00007fffffffda60 in ?? ()
#4  0x0000000803566810 in ?? ()
#5  0x0000000803868600 in ?? ()
#6  0x0000000000000000 in ?? ()
(gdb) quit
root@rspamd #

**Comment by vstakhov** [comment]
Could you enable debug logs for rspamd? I've observed something close today but that was caused by pcre-jit (the issue had disappeared after I've updated it to the recent version). To enable debug logging you might set `debug` instead of `info` in logging.conf.

**Comment by msimerson** [comment]
Gist of startup with [debugging enamed](https://gist.github.com/msimerson/12018dccbcf7541647dc)

**Comment by vstakhov** [comment]
And btw, AFAIR, gdb is terribly broken on FreeBSD 10 (I'm using merely `lldb` on this platform).

**Comment by msimerson** [comment]
[gist of a connection that failed](https://gist.github.com/msimerson/13d23334e4ed4ecf72e9)

> gdb is terribly broken on FreeBSD 10

Yeah, still building / installing lldb.

**Comment by vstakhov** [comment]
In this case, it would be also useful to grab a trace, to do it, you might run rspamd using `ktrace`:

```
# ktrace -i -d rspamd -u nobody -g nobody
^C after some time
# kdump
```

And similarly paste some lines near core dump (the output would be really huge).

**Comment by msimerson** [comment]
[ktrace output](http://matt.simerson.net/ktrace.txt)

**Comment by msimerson** [comment]
More logs (level=info):
````
2015-05-26 15:14:53 #18333(main) print_signals_info: got SIGCHLD from child: 18342; reason: 'Child has terminated abnormally but did not create a core file'
2015-05-26 15:14:53 #18333(main) main: normal process 18342 terminated abnormally by signal: 10
2015-05-26 15:14:55 #18333(main) print_signals_info: got signal: 'Alarm clock'; received from pid: 0; uid: 0
2015-05-26 15:14:55 #22613(normal) fork_worker: starting normal process 22613
2015-05-26 15:14:55 #22613(normal) rspamd_hosts_fin: read hash of 214 elements
2015-05-26 15:15:19 #18341(normal) accept_socket: accepted connection from 127.0.0.30 port 58126
2015-05-26 15:15:19 #18333(main) print_signals_info: got SIGCHLD from child: 18341; reason: 'Child has terminated abnormally but did not create a core file'
2015-05-26 15:15:19 #18333(main) main: normal process 18341 terminated abnormally by signal: 10
2015-05-26 15:15:21 #18333(main) print_signals_info: got signal: 'Alarm clock'; received from pid: 0; uid: 0
2015-05-26 15:15:21 #25099(normal) fork_worker: starting normal process 25099
2015-05-26 15:15:21 #25099(normal) rspamd_hosts_fin: read hash of 214 elements
2015-05-26 15:15:30 #18347(normal) accept_socket: accepted connection from 127.0.0.30 port 58140
2015-05-26 15:15:30 #18333(main) print_signals_info: got SIGCHLD from child: 18347; reason: 'Child has terminated abnormally but did not create a core file'
2015-05-26 15:15:30 #18333(main) main: normal process 18347 terminated abnormally by signal: 10
2
````

**Comment by msimerson** [comment]
> I've observed something close today but that was caused by pcre-jit

The recent pcre vulnerability is why I updated in the first place. 

Also, core files aren't being saved in the logs immediately above because I set `sysctl kern.coredump=0` above, as I've got plenty of core files already.

**Comment by vstakhov** [comment]
In your ktrace I don't see any crashes, in your debug output I don't see any scan attempts and in your gdb backtrace I don't see any useful trace. That makes me nuts :(

**Comment by msimerson** [comment]
Mmm, I may need to ktrace for longer.

**Comment by msimerson** [comment]
In the [debug gist](https://gist.githubusercontent.com/msimerson/13d23334e4ed4ecf72e9/raw/gistfile1.txt) is lots of stuff ending with this:
````
2015-05-26 14:28:15 #80020(main) print_signals_info: got SIGCHLD from child: 80097; reason: 'Child has terminated abnormally and created a core file'
````

Not what you want/need?

**Comment by vstakhov** [comment]
From your words, rspamd dies when you try to scan message. But I don't see any debug or even info level messages that are related to messages scanning (something like `accept_socket: accepted connection from 127.0.0.1 port xxxx`). Moreover, I don't see anything in the log file related to the workers. That's very confusing.

**Comment by msimerson** [comment]
Okay, very very much [ktrace output](http://matt.simerson.net/ktrace.txt) and lots and lots of [rspamd.log](http://matt.simerson.net/ktrace.txt), some of which is at level=debug.

**Comment by msimerson** [comment]
> rspamd dies when you try to scan message.

I don't recall saying such a thing. I recall posting log messages. Rspamd remains up and running, despite the premature death of child processes. The output of [rspamd.log](http://matt.simerson.net/rspamd.log) should reflect processes that look more or less like this:

````
root@rspamd:/ #	ps ax
  PID TT  STAT    TIME COMMAND
78820  -  IsJ  0:00.00 /usr/sbin/syslogd -ss
78883  -  IsJ  0:00.08 rspamd: main process (rspamd-0.9.4)
78891  -  IJ   0:00.00 rspamd: controller process (rspamd-0.9.4)
78894  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
78895  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
78899  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
78900  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
78902  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
78906  -  SJ   0:00.03 rspamd: normal process (rspamd-0.9.4)
78910  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
78913  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
78916  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
78920  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
78922  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
78923  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
82610  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
82628  -  SJ   0:00.02 rspamd: normal process (rspamd-0.9.4)
82706  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
82738  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
85581  6  SJ   0:00.01 su
85582  6  SJ   0:00.01 _su (sh)
85799  6  R+J  0:00.00 ps ax
root@rspamd:/ #
````

**Comment by vstakhov** [comment]
A quick question: what happens if you'd set `count = 1` in the worker section for normal worker?

**Comment by vstakhov** [comment]
And the second one: does rspamd listen on some socket?

**Comment by vstakhov** [comment]
Ah, well, I see that it listens.

**Comment by vstakhov** [comment]
Also maybe you might try this small patch and check what happens: http://highsecure.ru/test.diff

**Comment by msimerson** [comment]
> what happens if you'd set count = 1 in the worker section for normal worker

I get only one of the 'rspamd: normal process' processes.  With `count=1` and the patch manually applied, I still get these errors:
````
2015-05-26 19:40:03 #43160(normal) accept_socket: accepted connection from 127.0.0.30 port 48364
2015-05-26 19:40:03 #9567(main) print_signals_info: got SIGCHLD from child: 43160; reason: 'Child has terminated abnormally but did not create a core file'
````

> does rspamd listen on some socket

Several:

````
 # sockstat | grep :
nobody   rspamd-0.9 46851 8  tcp4   127.0.0.40:11334      *:*
nobody   rspamd-0.9 46851 9  tcp4   127.0.0.40:11333      *:*
nobody   rspamd-0.9 46851 10 udp4   127.0.0.40:19947      127.0.0.3:53
nobody   rspamd-0.9 46851 11 udp4   127.0.0.40:31127      127.0.0.3:53
nobody   rspamd-0.9 46851 12 udp4   127.0.0.40:56137      127.0.0.3:53
nobody   rspamd-0.9 46851 13 udp4   127.0.0.40:39216      127.0.0.3:53
nobody   rspamd-0.9 46851 14 udp4   127.0.0.40:33668      127.0.0.3:53
nobody   rspamd-0.9 46851 15 udp4   127.0.0.40:48776      127.0.0.3:53
nobody   rspamd-0.9 46851 16 udp4   127.0.0.40:13815      127.0.0.3:53
nobody   rspamd-0.9 46851 17 udp4   127.0.0.40:49127      127.0.0.3:53
nobody   rspamd-0.9 46341 8  tcp4   127.0.0.40:11334      *:*
nobody   rspamd-0.9 46341 9  udp4   127.0.0.40:59360      127.0.0.3:53
nobody   rspamd-0.9 46341 10 udp4   127.0.0.40:29627      127.0.0.3:53
nobody   rspamd-0.9 46341 11 udp4   127.0.0.40:35988      127.0.0.3:53
nobody   rspamd-0.9 46341 12 udp4   127.0.0.40:17965      127.0.0.3:53
nobody   rspamd-0.9 46341 13 udp4   127.0.0.40:61598      127.0.0.3:53
nobody   rspamd-0.9 46341 14 udp4   127.0.0.40:58132      127.0.0.3:53
nobody   rspamd-0.9 46341 15 udp4   127.0.0.40:63799      127.0.0.3:53
nobody   rspamd-0.9 46341 16 udp4   127.0.0.40:39632      127.0.0.3:53
nobody   rspamd-0.9 46340 8  tcp4   127.0.0.40:11334      *:*
nobody   rspamd-0.9 46340 9  tcp4   127.0.0.40:11333      *:*
````

### ps ax
````
  PID TT  STAT    TIME COMMAND
50863  -  IsJ  0:00.02 rspamd: main process (rspamd-0.9.4)
50864  -  SJ   0:00.00 rspamd: controller process (rspamd-0.9.4)
50865  -  SJ   0:00.00 rspamd: normal process (rspamd-0.9.4)
````

**Comment by vstakhov** [comment]
What about the patch then?

**Comment by vstakhov** [comment]
Or wait, I got it reproduced.

**Comment by vstakhov** [comment]
Do you have some custom statistics enabled or are you using the configuration files merely from the rspamd distribution?

**Comment by vstakhov** [comment]
Could you try apply this patch:

https://github.com/vstakhov/rspamd/commit/efb52618131c2c895802dc60433974085963c8a3

It fixes the similar issue in my test case.

**Comment by msimerson** [comment]
I'll give that a try, hopefully tonight or tomorrow when I have time.

**Comment by msimerson** [comment]
ef55261 applied, it's not a "cure":
````
[rspamd] query failed: socket hang up
2015-05-28 22:39:37 #1451(main) print_signals_info: got SIGCHLD from child: 1454; reason: 'Child has terminated abnormally but did not create a core file'
````

**Comment by andrejzverev** [comment]
Can you archive all yours rspamd configs (/usr/local/etc/rspamd/*) and put it somewhere on http(URL)?

**Comment by msimerson** [comment]
@andrejzverev , it's all stock configs. The only things I've changed are:
* worker-controller.inc: password="$1$7......"
* worker-controller.inc: secure_ip="127.0.0.40"
* worker-normal.inc: count=1   (per @vstakhov suggestion above)
* logging.inc: level="info"

**Comment by msimerson** [comment]
Fixed in 0.95
